8 toggle complete unit tests

jmrah opened the pull request.

jmrah committed:
Added unit tests for toggling a task between complete/incomplete

#8

jmrah committed:
Renamed test/unit/todoer.ts -> test/unit/toggle_task.ts

#8

jmrah committed:
Added more tests and refactored tests

#8

jmrah committed:
Refactored extension to use core/toder

#8

jmrah commented on src/core/todoer.ts:
```diff
@@ -1,6 +1,32 @@
+const incompleteTodos = /^- \[ \] \s*(.*?)\s*$/gm;
+
+// Example: "- [x] ~~Go shopping~~"
+const completeTodos = /^- \[x\] \s*~~\s*(.*?)\s*~~\s*$/gm;
```
gnarly regex, eh?

jmrah commented:
Hey @Gander7, I'm leaving this PR open.  I'm hoping you could pay attention to the test cases, and see if they all make sense.  I'll try highlighting a few that I wasn't sure about by commenting on them.

jmrah commented on test/unit/toggle_complete.ts:
```diff
@@ -0,0 +1,137 @@
+      "[ ] Go shopping",
+      "- [] Go shopping",
+      "- [ ]Go shopping",
+      "- [x] ~nope~~",
```
All these "nope" cases lol.  I'm considering them "not a todo" because they don't have the correct ~~ ~~ on both sides.  What do you think?  Too strict, or OK?

Gander7 replied:
Messaged you on teams :)

jmrah commented on test/unit/toggle_task.ts:
```diff
@@ -0,0 +1,132 @@
+
+  describe("Should toggle completed tasks to non-tasks", () => {
+    // prettier-ignore
+    const theories = [
```
Does this functinoality make sense to you?  If you have a completed task, you can turn it back into a non-task?

Gander7 replied:
Maybe have to uncomplete it first? Once we add things like startTime and endTime, this could potentially wipe out stats.

jmrah commented on test/unit/toggle_task.ts:
```diff
@@ -0,0 +1,132 @@
+
+  describe("Should toggle incomplete tasks to non-tasks", () => {
+    // prettier-ignore
+    const theories = [
```
Ended up using the term theories :)

Gander7 commented on test/unit/toggle_complete.ts:
```diff
@@ -0,0 +1,137 @@
+      "- [ ]Go shopping",
+      "- [x] ~nope~~",
+      "- [x] ~~nope~",
+      "- [x] nope",
```
This should be toggle-able right?

Gander7 commented on test/unit/toggle_complete.ts:
```diff
@@ -0,0 +1,137 @@
+      [ "- [ ] ",                  "- [x] ~~~~"                ],
+    ];
+
+    for (const [input, expected] of theories) {
```
Put the loop in a function?

Gander7 reviewed (changes requested):
Please remove tilde's as we talked about :)

jmrah committed:
Code review

- Keeps dash if it already exists when creating a task
- Keeps [] if it already exists when creating a task
- Can't turn a completed task into a non-task
- Only ~~ on cancelled tasks

#8

jmrah commented on src/core/todoer.ts:
```diff
@@ -1,6 +1,51 @@
+[ \t]*(.*?)[ \t]*                          # Capture the task text, leaving off any leading or trailing whitespace
+$
+*/
+const completeTasks = /^(?![ \t]*-[ \t]*\[[ \t]*\][ \t]*~~[^~]+~~)([ \t]*)-[ \t]*\[[ \t]*x[ \t]*\][ \t]*(.*?)[ \t]*$/gm;
```
These regexes are so complex, and I'm totally open to simpler solution if one exists in the future.  I felt like I was on the "regex or nothgin!!!!" train.  Coudn't think of another way.  I'm curious to know how well they'll be maintainable.  

Anyways, they seem to work!

jmrah commented on test/unit/toggle_complete.ts:
```diff
@@ -0,0 +1,142 @@
+
+describe("Completing a task", () => {
+  describe("Should toggle an incomplete task to complete", () => {
+    // prettier-ignore
```
I was getting pretty frustrated with prettier.   Love hate relationship I guess.

I like it for things like automatically putting (or removing) semicolons, chaning double quotes to single quotes (or vice versa), automatically indending my lines and code blocks.  But I don't like that it reformats my strings , or arrays, etc.
